cpp/src/strings/regex/regex.inl

```diff
@@ -40,95 +33,101 @@ namespace detail {
-  int32_t reserved;
-  int2* ranges{};       // pair per instruction
-  int16_t* inst_ids{};  // one per instruction
-  u_char* mask{};       // bit per instruction
-
-  CUDF_HOST_DEVICE inline static int32_t data_size_for(int32_t insts)
+  /**
+   * @brief Compute the memory size for the state data.
+   */
+  constexpr inline static std::size_t data_size_for(int32_t insts)
   {
     return ((sizeof(ranges[0]) + sizeof(inst_ids[0])) * insts) + ((insts + 7) / 8);
   }
 
-  CUDF_HOST_DEVICE inline static int32_t alloc_size(int32_t insts)
+  /**
+   * @brief Compute the aligned memory allocation size.
+   */
+  constexpr inline static std::size_t alloc_size(int32_t insts)
   {
-    int32_t size = sizeof(relist);
-    size += data_size_for(insts);
-    size = ((size + 7) / 8) * 8;  // align it too
-    return size;
+    return cudf::util::round_up_unsafe<size_t>(data_size_for(insts) + sizeof(relist),
+                                               sizeof(ranges[0]));
   }
 
-  CUDF_HOST_DEVICE inline relist() {}
+  struct alignas(16) restate {
+    int2 range;
+    int32_t inst_id;
+    int32_t reserved;
+  };
 
-  CUDF_HOST_DEVICE inline relist(int16_t insts, u_char* data = nullptr) : listsize(insts)
+  __device__ __forceinline__ relist(int16_t insts, u_char* data = nullptr)
+    : masksize((insts + 7) / 8)
   {
     auto ptr = data == nullptr ? reinterpret_cast<u_char*>(this) + sizeof(relist) : data;
     ranges   = reinterpret_cast<int2*>(ptr);
-    ptr += listsize * sizeof(ranges[0]);
+    ptr += insts * sizeof(ranges[0]);
     inst_ids = reinterpret_cast<int16_t*>(ptr);
-    ptr += listsize * sizeof(inst_ids[0]);
+    ptr += insts * sizeof(inst_ids[0]);
     mask = ptr;
     reset();
   }
 
-  CUDF_HOST_DEVICE inline void reset()
+  __device__ __forceinline__ void reset()
   {
-    memset(mask, 0, (listsize + 7) / 8);
+    memset(mask, 0, masksize);
     size = 0;
   }
 
-  __device__ inline bool activate(int32_t i, int32_t begin, int32_t end)
+  __device__ __forceinline__ bool activate(int32_t id, int32_t begin, int32_t end)
   {
-    if (readMask(i)) return false;
-    writeMask(true, i);
-    inst_ids[size] = static_cast<int16_t>(i);
+    if (readMask(id)) { return false; }
+    writeMask(id);
+    inst_ids[size] = static_cast<int16_t>(id);
     ranges[size]   = int2{begin, end};
     ++size;
     return true;
   }
 
-  __device__ inline void writeMask(bool v, int32_t pos)
+  __device__ __forceinline__ restate get_state(int16_t idx) const
   {
-    u_char uc = 1 << (pos & 7);
-    if (v)
-      mask[pos >> 3] |= uc;
-    else
-      mask[pos >> 3] &= ~uc;
+    return restate{ranges[idx], inst_ids[idx]};
   }
 
-  __device__ inline bool readMask(int32_t pos)
+  __device__ __forceinline__ int16_t get_size() const { return size; }
+
+ private:
+  int16_t size{};
+  int16_t masksize;
+  int32_t reserved;
+  int2* ranges;       // pair per instruction
+  int16_t* inst_ids;  // one per instruction
+  u_char* mask;       // bit per instruction
```

Comment:
Can we constify any of these things? I don't think the pointed to data can be const, but can the pointers themselves be made const? 

You might try and throw some `__restrict__`s on the pointer types too since we're just trying optimizations and seeing what sticks.

Reply:
Mainly the pointers cannot be const because that would require they be set in the ctor member initializer. And computing the pointers is fairly complicated in the ctor (and more complicated in the next PR).

Reply:
I'd still try `__restrict__` on these and other pointers. Especially the `relist` ones. 

 https://cellperformance.beyond3d.com/articles/2006/05/demystifying-the-restrict-keyword.html

Reply:
I was able to add `__restrict__` to the `relist` members (`ranges, inst_ids, mask`) but no to the `relist*` pointers in the `reljunk` struct. I get a compile error there:
```
/cudf/cpp/src/strings/regex/regex.cuh(192): error: "restrict" is not allowed
```

Reply:
Oops. I put it in front. Looks like it should go after.